src/gensearch.jl

```diff
@@ -0,0 +1,336 @@
+    Z = intersect(R, setdiff(ancestors(g, union(X, Y, I), veto), X, Y))
+    if alt_test_dsep(g, X, Y, Z, veto)
+        return Z
+    else
+        return false
+    end
+end
+
+function closure(g, X, A, Z, veto)
+    return gensearch(g, X, (pe, ne, v, w) -> (w in A) && !veto(pe, ne, v, w) && !(((pe == RIGHT && ne == RIGHT) || pe == LEFT) && v in Z))
+end
+
+"""
+    find_min_dsep(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y), veto = no_veto)
+
+Find an inclusion minimal d-separator `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, i.e., one for which no subset is a d-separator, else return `false`.
+
+Follows the algorithmic approach proposed in https://arxiv.org/abs/1803.00116. 
+"""
+function find_min_dsep(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y), veto = no_veto)
+    A = ancestors(g, union(X, Y, I), veto)
+    Z = find_dsep(g, X, Y, I, R, veto)
+    Z == false && return false
+    ZX = union(intersect(Z, closure(g, X, A, Z, veto)), I)
+    return union(intersect(ZX, closure(g, Y, A, ZX, veto)), I)
+end
+
+function pcp(g, X, Y)
+    return intersect(setdiff(descendants(g, X, no_incoming(X)), X), ancestors(g, Y, no_outgoing(X))
+)
+end
+
+"""
+    find_covariate_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+Find a covariate adjustment set `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, else return `false`. 
+
+Follows the algorithmic approach proposed in https://arxiv.org/abs/1803.00116. 
+"""
+function find_covariate_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    PCPXY = pcp(g, X, Y)
+    Z = setdiff(intersect(ancestors(g, union(X, Y, I)), R), X, Y, descendants(g, PCPXY))
+    if issubset(I, Z) && alt_test_dsep(g, X, Y, Z, (pe, ne, v, w) -> v in X && w in PCPXY && ne == RIGHT)
+        return Z
+    else
+        return false
+    end
+end
+
+"""
+    find_backdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+Find a backdoor adjustment set `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, else return `false`. 
+
+Follows the algorithmic approach proposed in https://arxiv.org/abs/1803.00116. 
+"""
+function find_backdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    bdZ = setdiff(find_covariate_adjustment(g, X, Y, I, R), descendants(g, X, no_incoming(X)))
+    if issubset(I, bdZ) && alt_test_dsep(g, X, Y, bdZ, no_outgoing(X))
+        return bdZ
+    else
+        return false
+    end
+end
+
+"""
+    find_covariate_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+Find an inclusion minimal covariate adjustment set `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`, else return `false`. 
+
+Follows the algorithmic approach proposed in https://arxiv.org/abs/1803.00116.  
+"""
+# this is also minimal bd set!
```

Comment:
Move comment so that docstring is above function name?

Reply:
thanks, I fixed it